source/websocket.c

```diff
@@ -0,0 +1,908 @@
+            websocket->thread_data.current_outgoing_frame = AWS_CONTAINER_OF(node, struct outgoing_frame, node);
+
+            struct aws_websocket_frame frame = {
+                .fin = websocket->thread_data.current_outgoing_frame->def.fin,
+                .opcode = websocket->thread_data.current_outgoing_frame->def.opcode,
+                .payload_length = websocket->thread_data.current_outgoing_frame->def.payload_length,
+            };
+
+            /* RFC-6455 Section 5.3 Client-to-Server Masking
+             * Clients must mask payload with key derived from an unpredictable source of entropy. */
+            if (!websocket->is_server) {
+                frame.masked = true;
+                /* TODO: faster source of random (but still seeded by device_random) */
+                struct aws_byte_buf masking_key_buf = aws_byte_buf_from_empty_array(frame.masking_key, 4);
+                err = aws_device_random_buffer(&masking_key_buf);
+                if (err) {
+                    AWS_LOGF_ERROR(
+                        AWS_LS_HTTP_WEBSOCKET,
+                        "id=%p: Failed to derive masking key, error %d (%s).",
+                        (void *)websocket,
+                        aws_last_error(),
+                        aws_error_name(aws_last_error()));
+                    goto error;
+                }
+            }
+
+            err = aws_websocket_encoder_start_frame(&websocket->thread_data.encoder, &frame);
+            if (err) {
+                AWS_LOGF_ERROR(
+                    AWS_LS_HTTP_WEBSOCKET,
+                    "id=%p: Failed to start frame encoding, error %d (%s).",
+                    (void *)websocket,
+                    aws_last_error(),
+                    aws_error_name(aws_last_error()));
+                goto error;
+            }
+
+            AWS_LOGF_TRACE(
+                AWS_LS_HTTP_WEBSOCKET,
+                "id=%p: Start writing frame=%p opcode=%" PRIu8 "(%s) payload-length=%" PRIu64 ".",
+                (void *)websocket,
+                (void *)websocket->thread_data.current_outgoing_frame,
+                websocket->thread_data.current_outgoing_frame->def.opcode,
+                aws_websocket_opcode_str(websocket->thread_data.current_outgoing_frame->def.opcode),
+                websocket->thread_data.current_outgoing_frame->def.payload_length);
+        }
+
+        err = aws_websocket_encoder_process(&websocket->thread_data.encoder, &io_msg->message_data);
+        if (err) {
+            AWS_LOGF_ERROR(
+                AWS_LS_HTTP_WEBSOCKET,
+                "id=%p: Frame encoding failed with error %d (%s).",
+                (void *)websocket,
+                aws_last_error(),
+                aws_error_name(aws_last_error()));
+            goto error;
+        }
+
+        if (aws_websocket_encoder_is_frame_in_progress(&websocket->thread_data.encoder)) {
+            AWS_LOGF_TRACE(
```

Comment:
maybe DEBUG on this one

Reply:
protip for commenting: Github shows 4 lines, the line you clicked on, plus the 3 above it. If you clicked on the string 2 lines down I'd have a better idea which log statement this was

Reply:
disagree on this one, this is an implementation detail about a frame of websocket data getting split across 2 aws_io_messages